relay-server/src/services/processor/groups/base.rs

```diff
@@ -0,0 +1,156 @@
+use std::sync::Arc;
+
+use relay_base_schema::project::ProjectId;
+use relay_event_schema::protocol::Event;
+use relay_quotas::RateLimits;
+
+use crate::services::processor::{InnerProcessor, ProcessingError, ProcessingExtractedMetrics};
+use crate::services::projects::project::ProjectInfo;
+use crate::utils::{ManagedEnvelope, TypedEnvelope};
+
+/// A macro that creates a new processing group type and implements necessary traits.
+/// This macro reduces boilerplate code for creating new processing group types by automatically
+/// implementing the group trait and conversion traits to/from processing group.
+#[macro_export]
+macro_rules! group {
+    ($variant:ident, $ty:ident) => {
+        #[derive(Clone, Copy, Debug)]
+        pub struct $ty;
+
+        impl Group for $ty {}
+
+        impl From<$ty> for ProcessingGroup {
+            fn from(_: $ty) -> Self {
+                ProcessingGroup::$variant
+            }
+        }
+
+        impl TryFrom<ProcessingGroup> for $ty {
+            type Error = GroupTypeError;
+
+            fn try_from(value: ProcessingGroup) -> Result<Self, Self::Error> {
+                if matches!(value, ProcessingGroup::$variant) {
+                    return Ok($ty);
+                }
+                return Err(GroupTypeError);
+            }
+        }
+    };
+}
+
+/// A macro that creates a function to process specific group types.
+/// Takes pairs of (processing group, process group) identifiers and generates match arms for
+/// each.
+#[macro_export]
+macro_rules! build_process_group {
+    ($(($group:ident, $process_group:ident)),*) => {
+        pub fn supports_new_processing(group: &ProcessingGroup) -> bool {
+            $(matches!(group, ProcessingGroup::$group) ||)* false
+        }
+
+        pub fn process_group(
+            group: ProcessingGroup,
+            managed_envelope: ManagedEnvelope,
+            processor: Arc<InnerProcessor>,
+            project_info: Arc<ProjectInfo>,
+            project_id: ProjectId,
+            rate_limits: Arc<RateLimits>,
+        ) -> Result<ProcessingResult, ProcessingError> {
+                match group {
+                    $(
+                        ProcessingGroup::$group => {
```

Comment:
I feel like it should be possible to do this dispatch directly with dynamic trait objects instead of an enum + macro.